### PR TITLE
Support both directory and block device for local volume plugin FileSystem VolumeMode

### PR DESCRIPTION
**Pull request by NickrenREN**
Support both directory and block device for local volume plugin FileSystem VolumeMode 

xref: [local storage dynamic provisioning design #1914](https://github.com/kubernetes/community/pull/1914)

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
Support both directory and block device for local volume plugin FileSystem VolumeMode 
```

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -1442,6 +1442,12 @@ type LocalVolumeSource struct {
 	// Block devices can be represented only by VolumeMode=Block, which also requires the
 	// BlockVolume alpha feature gate to be enabled.
 	Path string
+
+	// Filesystem type to mount.
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs". Implicitly inferred to be "ext4" if unspecified.
+	// +optional
+	FSType string
```
Comment (lichuqiang) [inline pkg/apis/core/types.go]: should be a pointer, as the field is optional

Comment (NickrenREN) [reply pkg/apis/core/types.go]: Other persistent volume sources are just using string, see https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L578
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L960
...

Comment (msau42) [reply pkg/apis/core/types.go]: New optional fields to existing types need to be pointers for backwards compatibility.  See https://github.com/kubernetes/kubernetes/issues/50121

File: pkg/volume/local/local.go [hunk 2]
```diff
@@ -205,6 +205,144 @@ func (plugin *localVolumePlugin) ConstructBlockVolumeSpec(podUID types.UID, volu
...
+		return devicePath, "", fmt.Errorf("local volume source is nil or local path is not set")
+	}
+	fileType, err := dm.mounter.GetFileType(spec.PersistentVolume.Spec.Local.Path)
+	if err != nil {
+		return devicePath, "", err
+	}
+
+	switch fileType {
+	case mount.FileTypeBlockDev:
+		return dm.mountLocalBlockDevice(spec, devicePath)
```
Comment (msau42) [inline pkg/volume/local/local.go]: Does `SetUp` also have to change, to bind mount from this global path instead?

File: pkg/volume/local/local.go [hunk 3]
```diff
@@ -205,6 +205,144 @@ func (plugin *localVolumePlugin) ConstructBlockVolumeSpec(podUID types.UID, volu
...
+	case mount.FileTypeDirectory:
+		// if the given local volume path is of already filesystem directory, return directly
+		return spec.PersistentVolume.Spec.Local.Path, spec.PersistentVolume.Spec.Local.Path, nil
```
Comment (msau42) [inline pkg/volume/local/local.go]: I wonder, can we return nil path for this case?  Could UnmountDevice be simplified if it is only called for blockdev scenario?

File: pkg/apis/core/types.go [hunk 4]
```diff
@@ -1442,6 +1442,12 @@ type LocalVolumeSource struct {
 	// Block devices can be represented only by VolumeMode=Block, which also requires the
```
Comment (msau42) [inline pkg/apis/core/types.go]: This comment also needs to be updated because now you can specify a block device with VolumeMode=Filesystem.

File: pkg/apis/core/types.go (hunk 1 above)
Comment (msau42) [reply pkg/apis/core/types.go]: ah sorry, I think we may not be able to default the field, because it would not work on windows.

File: pkg/volume/local/local.go (hunk 2 above)
Comment (NickrenREN) [reply pkg/volume/local/local.go]: yes. i think so

File: pkg/volume/local/local.go (hunk 3 above)
Comment (NickrenREN) [reply pkg/volume/local/local.go]: I think they are the same. deviceMountPath is set to local pv path If it is filesystem dir. and UnmountPath will not be executed for this case.

File: pkg/apis/core/types.go (hunk 4 above)
Comment (NickrenREN) [reply pkg/apis/core/types.go]: Deleted the last three lines

File: pkg/volume/local/local.go [hunk 5]
```diff
@@ -205,6 +210,150 @@ func (plugin *localVolumePlugin) ConstructBlockVolumeSpec(podUID types.UID, volu
...
+			}
+			notMnt = true
+		} else {
+			return devicePath, deviceMountPath, err
+		}
+	}
+	if !notMnt {
+		return devicePath, deviceMountPath, nil
+	}
+	fstype, err := getVolumeSourceFSType(spec)
```
Comment (msau42) [inline pkg/volume/local/local.go]: Do we need to default fstype to ext4?

Comment (NickrenREN) [reply pkg/volume/local/local.go]: `FormatAndMount` will do it if we didn't do the default, but yeah, we can do this here explicitly

**Comment by msau42** [comment]
Regarding the test failure with symlinks, I think it's because of the protected symlinks feature, which disallows following symlinks in world writable directories if the owner of the directory + link doesn't match the user trying to access it.  So I think you can try creating the directory + symlink with sudo.  [ref](https://stackoverflow.com/questions/26496352/symlink-giving-permission-denied-to-root)

**Comment by jsafrane** [comment]
@msau42 @NickrenREN, was there any conclusion about this PR? I am still interested in it. I don't want to mount local devices before consuming them in pods as filesystems.

**Comment by msau42** [comment]
This PR depends on attacher interface refactor which has not made progress in discussion

**Comment by NickrenREN** [comment]
yes, this PR is based on attacher refactor PR, once that one is done, i will rebase this PR...

**Comment by NickrenREN** [comment]
@msau42 @jsafrane  Since the [refactor PR](https://github.com/kubernetes/kubernetes/pull/66884) gets merged, this one can go now.
It is ready for review. Although i may need to add more UTs
PTAL when you get a chance, thanks

File: pkg/volume/local/local.go [hunk 6]
```diff
@@ -207,6 +216,156 @@ func (plugin *localVolumePlugin) ConstructBlockVolumeSpec(podUID types.UID, volu
...
+	}
+}
+
+func getVolumeSourceFSType(spec *volume.Spec) (string, error) {
+	if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.Local != nil {
+		if spec.PersistentVolume.Spec.Local.FSType != nil {
+			return *spec.PersistentVolume.Spec.Local.FSType, nil
+		} else {
+			return "", nil
```
Comment (msau42) [inline pkg/volume/local/local.go]: Set the default here instead of in the caller?

File: pkg/volume/local/local.go [hunk 7]
```diff
@@ -207,6 +216,156 @@ func (plugin *localVolumePlugin) ConstructBlockVolumeSpec(podUID types.UID, volu
...
+func (plugin *localVolumePlugin) GetDeviceMountRefs(deviceMountPath string) ([]string, error) {
+	mounter := plugin.host.GetMounter(plugin.GetPluginName())
+	return mounter.GetMountRefs(deviceMountPath)
+}
+
+var _ volume.DeviceUnmounter = &deviceMounter{}
+
+func (dm *deviceMounter) UnmountDevice(deviceMountPath string) error {
+	basemountPath := filepath.Join(dm.plugin.host.GetPluginDir(localVolumePluginName), mount.MountsInGlobalPDPath)
+	if strings.Contains(deviceMountPath, basemountPath) {
```
Comment (msau42) [inline pkg/volume/local/local.go]: better to use `mount.pathWithinBase` to properly handle substrings.  (this function may need to be exported in the Mount interface)

File: pkg/volume/local/local.go [hunk 8]
```diff
@@ -205,6 +205,139 @@ func (plugin *localVolumePlugin) ConstructBlockVolumeSpec(podUID types.UID, volu
...
+
+func (plugin *localVolumePlugin) GetDeviceMountRefs(deviceMountPath string) ([]string, error) {
+	mounter := plugin.host.GetMounter(plugin.GetPluginName())
+	return mount.GetMountRefs(mounter, deviceMountPath)
+}
+
+var _ volume.DeviceUmounter = &deviceMounter{}
+
+func (dm *deviceMounter) UnmountDevice(deviceMountPath string) error {
+	basemountPath := filepath.Join(dm.plugin.host.GetPluginDir(localVolumePluginName), mount.MountsInGlobalPDPath)
```
[the remark this answers is not in the document]
Comment (msau42) [reply pkg/volume/local/local.go]: Can you add a comment here explaining the `deviceMountPath` formats that can be passed into here?  That will help better understand why we can't just call `Dir()`.

File: pkg/volume/local/local.go [hunk 9]
```diff
@@ -207,6 +216,156 @@ func (plugin *localVolumePlugin) ConstructBlockVolumeSpec(podUID types.UID, volu
...
+
+	fileType, err := plugin.host.GetMounter(plugin.GetPluginName()).GetFileType(spec.PersistentVolume.Spec.Local.Path)
+	if err != nil {
+		return "", err
+	}
+	switch fileType {
+	case mount.FileTypeDirectory:
+		return spec.PersistentVolume.Spec.Local.Path, nil
+	case mount.FileTypeBlockDev:
+		return filepath.Join(plugin.host.GetPluginDir(localVolumePluginName), mount.MountsInGlobalPDPath, spec.Name()), nil
```
Comment (msau42) [inline pkg/volume/local/local.go]: Can this directory construction be a helper function since it's used in a few places?

Also, MountsInGlobalPDPath is a strange variable name, maybe it should be renamed in a separate PR?

**Review by msau42 — COMMENTED** [review]
There should be unit and e2e tests that use Local.Path = block with PV.volumeMode = Filesystem.

File: pkg/volume/local/local.go (hunk 6 above)
Comment (NickrenREN) [reply pkg/volume/local/local.go]: done

File: pkg/volume/local/local.go (hunk 7 above)
Comment (NickrenREN) [reply pkg/volume/local/local.go]: done

File: pkg/volume/local/local.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (NickrenREN) [reply pkg/volume/local/local.go]: sure, done

File: pkg/apis/core/types.go [hunk 10]
```diff
@@ -1499,10 +1499,13 @@ type KeyToPath struct {
...
 	// It can be either a directory or block device (disk, partition, ...).
-	// Directories can be represented only by PersistentVolume with VolumeMode=Filesystem.
-	// Block devices can be represented only by VolumeMode=Block, which also requires the
-	// BlockVolume alpha feature gate to be enabled.
 	Path string
+
+	// Filesystem type to mount.
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs". Implicitly inferred to be "ext4" if unspecified.
+	// +optional
```
Comment (jsafrane) [inline pkg/apis/core/types.go]: Please add some note that it applies only when the Path is a block device.

File: test/e2e/storage/persistent_volumes-local.go [hunk 11]
```diff
@@ -1038,15 +1038,12 @@ func setupLocalVolumeBlockFs(config *localTestConfig, node *v1.Node) *localTestV
...
-	// give others rwx for read/write testing
-	cmd := fmt.Sprintf("sudo mkfs -t ext4 %s && sudo mount -t ext4 %s %s && sudo chmod o+rwx %s", loopDev, loopDev, hostDir, hostDir)
-	_, err := issueNodeCommandWithResult(config, cmd, node)
-	Expect(err).NotTo(HaveOccurred())
 	// Populate block volume with testFile containing testFileContent.
-	volume := setupWriteTestFile(hostDir, config, BlockFsLocalVolumeType, node)
-	volume.hostDir = hostDir
-	volume.loopDevDir = loopDev
+	volume := setupWriteTestFile(loopDev, config, BlockFsLocalVolumeType, node)
+	// we do this in order to set block device path to local PV spec path directly
```
Comment (msau42) [inline test/e2e/storage/persistent_volumes-local.go]: Instead of changing the existing test, I would add a new volume type.  So we basically have 3 cases for a block device:
* Path = block, VolumeMode = block
* Path = block, VolumeMode = fs
* Path = fs, VolumeMode = fs

Comment (NickrenREN) [reply test/e2e/storage/persistent_volumes-local.go]: done

File: test/e2e/storage/persistent_volumes-local.go [hunk 12]
```diff
@@ -1338,32 +1367,46 @@ func unmapBlockLocalVolume(config *localTestConfig, dir string, node *v1.Node) {
...
-		sudoCmd := fmt.Sprintf("SUDO_CMD=$(which sudo); echo ${SUDO_CMD}")
-		// Write the testFileContent into the block device.
-		writeBlockCmd := fmt.Sprintf("${SUDO_CMD} dd if=%s of=%s bs=512 count=100", testFilePath, testDir)
-		// Cleanup the file containing testFileContent.
-		deleteTestFileCmd := fmt.Sprintf("rm %s", testFilePath)
-		return fmt.Sprintf("%s && %s && %s && %s", writeTestFileCmd, sudoCmd, writeBlockCmd, deleteTestFileCmd)
+func createWriteCmd(testDir string, testFile string, writeTestFileContent string, volumeType localVolumeType, whenSetUp bool) string {
+	if whenSetUp {
+		// when setup, we need to test block device command if volumeType is BlockLocalVolumeType or BlockFsWithoutFormatLocalVolumeType
+		if volumeType == BlockLocalVolumeType || volumeType == BlockFsWithoutFormatLocalVolumeType {
```
Comment (msau42) [inline test/e2e/storage/persistent_volumes-local.go]: I'm not sure if we need to treat setup specially.  Setup should create and format a fs, and then we can write a file to it just like any other fs case.

Comment (NickrenREN) [reply test/e2e/storage/persistent_volumes-local.go]: Setup will not format block device for the new test case (do not format it manually, kubernetes will do when mounting), so when setup, we should treat it as block device, then when we create a pod, referencing it, the block device will be formatted to a specific fs type. So we should test that differently.

Comment (msau42) [reply test/e2e/storage/persistent_volumes-local.go]: Ah I see.  The read/write test cases currently expect the content to be written during setup time.

Maybe it would be better to restructure those tests to not pre-populate data during setup, and instead in the BeforeEach(), create a Pod that mounts the local PV and writes data to it.

I think that would be cleaner rather then "skipping" the test by returning success for this case, or writing a whole different set of tests just for this type.

Comment (NickrenREN) [reply test/e2e/storage/persistent_volumes-local.go]: We will not `skipping` all the tests for the new test case
Each local volume type has several e2e test cases ,and we can already test the condition you mentioned above.
For example, in `twoPodsReadWriteTest`, we create two pods, and try to read file content immediately, but since the new test case will be treated as block device in setup, here, we will skip the reading test.  then, we will run `createWriteCmd`, pod is created now, we will write the file as we do for fs type(`mkdir, and echo filecontent to testfile`), and later `testReadFileContent` will read the file content then.
I think the test above will test all the conditions for the new test case well.
- when pod is not created(in setup), the local volume is a block device
- then, when pod is created, it will be formatted by kubernetes, we can treat it as a FS type

**Comment by dims** [comment]
@NickrenREN @saad-ali is this targetted for v1.12? Please set the milestone if so.

File: test/e2e/storage/persistent_volumes-local.go (hunk 12 above)
Comment (msau42) [reply test/e2e/storage/persistent_volumes-local.go]: Yes, the read/write will still be executed in some subset of the test cases.  I think it will be cleaner though if all test cases behave the same way across all types.

**Comment by NickrenREN** [comment]
@dims yes
could you help set the milestone? i can not do that. @dims @saad-ali

File: test/e2e/storage/persistent_volumes-local.go (hunk 12 above)
Comment (NickrenREN) [reply test/e2e/storage/persistent_volumes-local.go]: get it, and updated

**Comment by msau42** [comment]
Thanks for working on this!

File: pkg/volume/local/local.go [hunk 13]
```diff
@@ -38,6 +38,10 @@ import (
 	"k8s.io/kubernetes/pkg/volume/validation"
 )
 
+const (
+	defaultFSType = "ext4"
```
Comment (thockin) [inline pkg/volume/local/local.go]: Why do we need to make this assumption?  In general Linux can auto-detect filesystems, right?  why not make the default be best-effort auto-detect, and try to avoid encouraging people to make assumptions.

This sort of stuff is also hard to ferret out as Windows support becomes more of a Real Thing (@kubernetes/sig-windows-misc )

```
# cd /tmp

# dd if=/dev/zero of=loopfile bs=$((1024*1024)) count=64
64+0 records in
64+0 records out
67108864 bytes (67 MB, 64 MiB) copied, 0.0661057 s, 1.0 GB/s

# mkfs.xfs ./loopfile 
meta-data=./loopfile             isize=512    agcount=4, agsize=4096 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=0, rmapbt=0, reflink=0
data     =                       bsize=4096   blocks=16384, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0 ftype=1
log      =internal log           bsize=4096   blocks=855, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0

# mkdir mnt

# mount -oloop ./loopfile ./mnt

# mount | grep loopfile
/tmp/loopfile on /tmp/mnt type xfs (rw,relatime,attr2,inode64,noquota)

# umount ./mnt
```

Comment (thockin) [reply pkg/volume/local/local.go]: For the case of format-on-demand, (assuming that's what really drives this) perhaps it should be "whatever the node in question thinks is best"?  Can we make that something the host defines?  I feel like we've had this conversation before and I lost it then, but I forget why, so I'll happily take another tilt at it..

Comment (msau42) [reply pkg/volume/local/local.go]: The primary purpose of this field is for the format-on-demand case, not really for identifying the filesystem type on existing volumes.

We could remove the fstype field and just format it with "whatever the host wants", but how can we support situations where the user wants something different?

**Comment by thockin** [comment]
Leaving the API for the user to specify it (even though I really dislike
that still) but make the node determine which FS to use if not specified.
HOW to make the node specify that is what I don't know..

**Comment by msau42** [comment]
> make the node determine which FS to use if not specified.  HOW to make the node specify that is what I don't know..

What about an OS check for now?  If Linux, use ext4, if Windows, use ntfs.

**Comment by thockin** [comment]
The goal is not to spell "ext4" anywhere in Kubernetes.  Just as we want to
never spell "google" or "amazon" and move all of that out of core, we
should endeavour to reduce hard deps on linuxisms, especially ones that
change vogue over time or across sites.

E.g. `func (mounter *SafeFormatAndMount) formatAndMount` should not
hardcode ext4 and flags for it, but should either exec something from the
node itself, or should receive params (e.g. kubelet config params) that
specify which FS and flags to use by default, or maybe even more flexibly
which FSes are allowed, which FS is default, and a map of per-fs flags to
use for mkfs and mount.

Then we can delete hardcoded "ext4" stuff and move it to configs, where it
can be changed more readily.

Thoughts?  I could be convinced to allow this if I knew that the larger
work was at least rough-designed and planned (if not staffed).

**Comment by msau42** [comment]
Ok I see what you mean.  Long term, CSI still delegates the formatting operation to the CSI plugin and leaves the decision of which filesystem to use up to the plugin.  This doesn't solve the issue where the host doesn't have the appropriate kernel modules for a filesystem.  

One way this could be solved is to have the admin specify a default filesystem to the CSI driver when deploying it.  

Another, more automated way is along the lines of what you are describing, to have some Node level config specifying the default and supported filesystems, and have the CSI Kubernetes adapter pass that to the CSI driver.  This method has not been designed or scoped out.  Do you think this is absolutely necessary, or can we live with having an admin/provider configure the CSI plugin?

**Comment by thockin** [comment]
Admin configured is acceptable to me

**Comment by NickrenREN** [comment]
auto-generated files conflict, and rebased

@msau42 @thockin do we need to create an issue to track the fstype setting ? It seems that we should not do that in this one.

**Comment by thockin** [comment]
I just want an issue filed or something, tagged appropriately, capturing what we talked about and linking back ti this PR.

**Comment by NickrenREN** [comment]
@thockin  issue created: https://github.com/kubernetes/kubernetes/issues/67834

File: pkg/apis/core/types.go [hunk 14]
```diff
@@ -1499,10 +1499,14 @@ type KeyToPath struct {
...
 	// It can be either a directory or block device (disk, partition, ...).
-	// Directories can be represented only by PersistentVolume with VolumeMode=Filesystem.
-	// Block devices can be represented only by VolumeMode=Block, which also requires the
-	// BlockVolume alpha feature gate to be enabled.
 	Path string
+
+	// Filesystem type to mount.
+	// It applies only when the Path is a block device.
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs". Implicitly inferred to be "ext4" if unspecified.
```
Comment (thockin) [inline pkg/apis/core/types.go]: Please change the documentation here (and v1)  to say that the default value is to auto-select a fileystem, rather than saying it is ext4.  That way we can actually change it later.

Comment (NickrenREN) [reply pkg/apis/core/types.go]: should this be changed when we actually chane it ? Since now it still defaults to 'ext4'.
btw: some other plugins are saying like this too

Comment (msau42) [reply pkg/apis/core/types.go]: Let's change it now so we don't have to change it again later.  (and the new wording is still true to the implementation)

Comment (NickrenREN) [reply pkg/apis/core/types.go]: ok, updated

**Comment by NickrenREN** [comment]
@msau42 @thockin any comments ?

**Comment by msau42** [comment]
/lgtm
ping @thockin for approval

**Comment by NickrenREN** [comment]
failing tests are not related to this change , issue created: https://github.com/kubernetes/kubernetes/issues/68006

**Comment by PatrickLang** [comment]
@KnicKnic - can you review from Windows standpoint and help build a plan for how to hook this up? I think block devices may be feasible in Windows Server 2019 with some changes to ContainerD. 

@andyzhangx - can you review from Azure standpoint?

**Comment by msau42** [comment]
@NickrenREN can you also investigate if the functionality of local volumes (bind mount/read/write/fsgroup) would have worked before this change if the volume was created with permissions that kubelet did not have?

We need to figure out if this has the potential to be a regression, and if so, we need an appropriate release note with "ACTION REQUIRED"

**Comment by msau42** [comment]
We discussed, this is only an issue with /tmp directories, which is only a testing issue
